flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java

```diff
@@ -0,0 +1,258 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.common.typeutils;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Base class for composite serializers.
+ *
+ * <p>This class serializes a composite type using array of its field serializers.
+ * Fields are indexed the same way as their serializers.
+ *
+ * @param <T> type of custom serialized value
+ */
+public abstract class CompositeSerializer<T> extends TypeSerializer<T> {
+	private static final long serialVersionUID = 1L;
+
+	protected final TypeSerializer<Object>[] fieldSerializers;
+	final boolean isImmutableTargetType;
+	private final int length;
+
+	@SuppressWarnings("unchecked")
+	protected CompositeSerializer(boolean isImmutableTargetType, TypeSerializer<?> ... fieldSerializers) {
+		Preconditions.checkNotNull(fieldSerializers);
+		Preconditions.checkArgument(Arrays.stream(fieldSerializers).allMatch(Objects::nonNull));
+		this.isImmutableTargetType = isImmutableTargetType;
+		this.fieldSerializers = (TypeSerializer<Object>[]) fieldSerializers;
+		this.length = calcLength();
+	}
+
+	/** Create new instance from its fields.  */
+	public abstract T createInstance(@Nonnull Object ... values);
+
+	/** Modify field of existing instance. Supported only by mutable types. */
+	protected abstract void setField(@Nonnull T value, int index, Object fieldValue);
+
+	/** Get field of existing instance. */
+	protected abstract Object getField(@Nonnull T value, int index);
+
+	/** Factory for concrete serializer. */
+	protected abstract CompositeSerializer<T> createSerializerInstance(TypeSerializer<?> ... originalSerializers);
+
+	@Override
+	public CompositeSerializer<T> duplicate() {
+		TypeSerializer[] duplicatedSerializers = new TypeSerializer[fieldSerializers.length];
+		boolean stateful = false;
+		for (int index = 0; index < fieldSerializers.length; index++) {
+			duplicatedSerializers[index] = fieldSerializers[index].duplicate();
+			if (fieldSerializers[index] != duplicatedSerializers[index]) {
+				stateful = true;
+			}
+		}
+		return stateful ? createSerializerInstance(duplicatedSerializers) : this;
+	}
+
+	@Override
+	public boolean isImmutableType() {
+		for (TypeSerializer<Object> fieldSerializer : fieldSerializers) {
```

Comment:
Why not compute this once in the constructor and remember in a boolean flag??

Reply:
We can compute many things like length, immutability, etc already in the constructor. Statelessness is the one thing that we might want to figure out and remember on the first attempt.